### PR TITLE
Delete invalid certs during tidy operation

### LINKED ISSUE #659: Tidy errors out on certificates with duplicate extension (Vault #27219)
This is an issue reported on vault [issues/27219](https://github.com/hashicorp/vault/issues/27219) by @Garagoth. 

**Description**
When a certificate with duplicate extensions exists and tidy is run, the process fails:
```
unable to parse stored certificate with serial \"06-e0-83-88-22-be-91-7f-8d-07-2e-21-35-01-f1-66-95-a0-de-35\": x509: certificate contains duplicate extensions
```

This is the line catching the failure: 
https://github.com/openbao/openbao/blob/c9201295ed833b431249f4592f32b1946b69f263/builtin/logical/pki/path_tidy.go#L875

PR [16700](https://github.com/hashicorp/vault/pull/16700) solved the problem of creating these invalid certificates but there's no way to get rid of them.

**To Reproduce**
1. Have a certificate with duplicate extensions. 
2. Run tidy process.

**Expected behavior**
Since certificates with duplicate extensions are deemed invalid in [RFC 5290, Section 4.2](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2), a way to delete these invalid certificates should exist.

If the tidy operation is run with certificates with extra extensions, they should be skipped OR the tidy operation stops and a warning which includes the command to delete them should be thrown. 

---
I'm working on the fix for this :)

### PR DESCRIPTION
**Pull request by fatima2003**
This PR addresses the issue where the tidy operation fails when certificates with duplicate extensions are present in storage as described by @Garagoth in [Vault Issue #27219](https://github.com/hashicorp/vault/issues/27219). Certificates that have duplicate X.509 extensions are deemed invalid according to [RFC 5280, Section 4.2](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2), and can no longer be parsed by the Go x509.ParseCertificate function since Go 1.19 (due to golang/go#50988).

Additionally, the [parseCertificate](https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/crypto/x509/parser.go;l=829) function in the Go library returns other errors when a certificate is invalid, such as when the issuer is malformed, &c.. Instead of only handling errors related to certificates with duplicate extensions, I suggest deleting any certificate that encounters parsing errors, as these certificates will be unusable anyway. The potential downside of this approach is that it would deem only certificates parsable by the official Go function as valid. However, this may not be an issue since that function is currently used for parsing. 

#### Potential Solutions:
1. Delete these certs during tidy. However, this risks that some certificates could be unintentionally deleted or worst case scenario, a bug with parsing could wipe out the cert store.

2. (**Currently implemented**) Have a separate flag for this task similar to `tidy_revoked_certs` called `tidy_invalid_certs` so tidying these certs could look like: 
```
bao write pki_int/tidy tidy_invalid_certs=true
```
This was suggested to eliminate the risk of deleting certs unintentionally.

3. Similar to option 2,  but instead of tidying all the invalid certs, the user gives the serial number of the invalid cert so they can pick which ones to delete:
```
bao write pki_int/tidy tidy_invalid_certs=[XXXXXXX, YYYYYYY]
```
Same benefit as option 2 but allows for finer control for cert deletion, though I'm not sure if invalid certs have a use case so maybe this is an overkill. 

#### TO DO 

- [X] Add flag to tidy invalid certs.
- [X]  Create test for this.
- [X] Add docs.

---
Resolves #659 [Vault #27219](https://github.com/hashicorp/vault/issues/27219)

## Target Release
1.14.7

### CODE REVIEW

**Comment by fatima2003** [comment]
I realized a flaw in the implementation yesterday... currently the `tidy_invalid_certs` flag cannot be called on its own since I purposely did not add it to `AnyTidyConfig` (to reduce the number of times all certs are iterated over) and the functionality only works when `tidy_cert_store` is also enabled. I think this design is confusing to the end user. I assume that people will not have to enable this flag often and for now I think only those with certificates with duplicate extensions will find this useful so I will make the changes so that it can be called on its own...

**Comment by cipherboy** [comment]
@fatima2003 Here's an invalid cert to test with, courtesy of the Golang test suite:

```
-----BEGIN CERTIFICATE-----
MIIBrjCCARegAwIBAgIBATANBgkqhkiG9w0BAQsFADAPMQ0wCwYDVQQDEwR0ZXN0
MCIYDzAwMDEwMTAxMDAwMDAwWhgPMDAwMTAxMDEwMDAwMDBaMA8xDTALBgNVBAMT
BHRlc3QwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAMiFchnHms9l9NninAIz
SkY9acwl9Bk2AtmJrNCenFpiA17AcOO5q8DJYwdXi6WPKlVgcyH+ysW8XMWkq+CP
yhtF/+LMzl9odaUF2iUy3vgTC5gxGLWH5URVssx21Und2Pm2f4xyou5IVxbS9dxy
jLvV9PEY9BIb0H+zFthjhihDAgMBAAGjFjAUMAgGAioDBAIFADAIBgIqAwQCBQAw
DQYJKoZIhvcNAQELBQADgYEAlhQ4TQQKIQ8GUyzGiN/75TCtQtjhMGemxc0cNgre
d9rmm4DjydH0t7/sMCB56lQrfhJNplguzsbjFW4l245KbNKHfLiqwEGUgZjBNKur
ot6qX/skahLtt0CNOaFIge75HVKe/69OrWQGdp18dkay/KS4Glu8YMKIjOhfrUi1
NZA=
-----END CERTIFICATE-----
```

The serial number is 1, but that shouldn't cause any issues.

File: changelog/665.txt
Comment (cipherboy) [inline changelog/665.txt:2]:
```suggestion
secret/pki: Delete invalid certificates during tidy if they cannot be parsed due to Go's x509 handling and `tidy_invalid_certs=true`.
```

File: builtin/logical/pki/path_tidy.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy.go:807]: I don't know that I like the idea of this being a separate pass. We already have one pass which does this (`config.CertStore` controlling `b.doTidyCertStore`), so I think it is fine to reuse that for this pass as well. In the case of 1M+ certs, we don't really want to iterate over all the certificates twice, especially if most certs aren't invalid. :-)

My 2c!

File: builtin/logical/pki/path_tidy.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy.go:909]: In particular, I'd make this logic something like:

```suggestion
			if !config.InvalidCerts && !haveWarned {
				logger.Warn("unable to parse stored certificate with serial %q: %w; tidy by enabling tidy_invalid_certs", serial, err)
				haveWarned = true
			} else {
				if err := req.Storage.Delete(ctx, "certs/"+serial); err != nil {
					return fmt.Errorf("error deleting invalid certificate %s: %w", serial, err)
				}
				b.tidyStatusIncCertStoreCount()
			}
```

or something. :D

**Review by cipherboy — COMMENTED** [review]
Some initial thoughts!

File: builtin/logical/pki/path_tidy.go
Comment (cipherboy) [reply builtin/logical/pki/path_tidy.go:807]: On line 784: `if config.CertStore || config.InvalidCerts {` ...

File: builtin/logical/pki/path_tidy.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy.go:910]: Ahh, creative! Just a note though, this isn't a `printf`-style logger, this is a `key=value` style logger with explicit message.

So I think we'd need something like the following:

```suggestion
				msg := "Unable to parse stored certificate. Other invalid certificates may exist; "
				if config.InvalidCerts {
					msg += "tidying up since it is not usable."
				} else {
					msg += "tidy by enabling tidy_invalid_certs=true."
				}
				logger.Warn(msg, "serial", serial, "err", err)
```

as otherwise you'll get weird logger messages. I believe this wouldn't err, but would output something like:

```
[WARN] msg="unable to parse stored certificate with serial %q: %w; %s. Other invalid certificates may exist."  1d:de:f6:bc:77:2e:56:cb:b7:00:bc:5b:ea:db:37:63:0a:3e:2d:98="x509: certificate contains duplicate extensions"
```

which wouldn't make as much sense.

File: changelog/665.txt
Comment (cipherboy) [inline changelog/665.txt:2]: nit:

```suggestion
secret/pki: Delete invalid certificates during tidy via `tidy_invalid_certs=true` if they cannot be parsed due to Go's x509 handling.
```

File: builtin/logical/pki/path_tidy.go
Comment (fatima2003) [reply builtin/logical/pki/path_tidy.go]: Yes, I initially also wanted it to be a single pass but figured since the `tidy_invalid_certs` would be rarely used it wouldn't matter so much but this is much better, thank you :D

File: builtin/logical/pki/path_tidy.go
Comment (fatima2003) [reply builtin/logical/pki/path_tidy.go]: I made the changes :D Thank you for the review + suggestions.

File: builtin/logical/pki/path_tidy.go
Comment (fatima2003) [reply builtin/logical/pki/path_tidy.go:910]: Ah interesting, I will keep this in mind for future logging :D Thank you

File: changelog/665.txt
Comment (fatima2003) [reply changelog/665.txt:2]: I'll add all these suggestions in one commit with the test :D Thank you!

**Comment by fatima2003** [comment]
> @fatima2003 Here's an invalid cert to test with, courtesy of the Golang test suite:
> 
> ```
> -----BEGIN CERTIFICATE-----
> MIIBrjCCARegAwIBAgIBATANBgkqhkiG9w0BAQsFADAPMQ0wCwYDVQQDEwR0ZXN0
> MCIYDzAwMDEwMTAxMDAwMDAwWhgPMDAwMTAxMDEwMDAwMDBaMA8xDTALBgNVBAMT
> BHRlc3QwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAMiFchnHms9l9NninAIz
> SkY9acwl9Bk2AtmJrNCenFpiA17AcOO5q8DJYwdXi6WPKlVgcyH+ysW8XMWkq+CP
> yhtF/+LMzl9odaUF2iUy3vgTC5gxGLWH5URVssx21Und2Pm2f4xyou5IVxbS9dxy
> jLvV9PEY9BIb0H+zFthjhihDAgMBAAGjFjAUMAgGAioDBAIFADAIBgIqAwQCBQAw
> DQYJKoZIhvcNAQELBQADgYEAlhQ4TQQKIQ8GUyzGiN/75TCtQtjhMGemxc0cNgre
> d9rmm4DjydH0t7/sMCB56lQrfhJNplguzsbjFW4l245KbNKHfLiqwEGUgZjBNKur
> ot6qX/skahLtt0CNOaFIge75HVKe/69OrWQGdp18dkay/KS4Glu8YMKIjOhfrUi1
> NZA=
> -----END CERTIFICATE-----
> ```
> 
> The serial number is 1, but that shouldn't cause any issues.

**Comment by cipherboy** [comment]
@fatima2003 Ah, do a direct `s.Put(...)` operation on an entry with a Key of certs/1 and a value which is the DER of the cert (you'll want to base64 decode it).

**Comment by fatima2003** [comment]
@cipherboy Oh okay, I add it like that :)

sos I accidentally clicked close with comment instead of comment 🙇🏽‍♀️

**Comment by fatima2003** [comment]
@cipherboy if all looks good, I can rebase :D

File: builtin/logical/pki/path_tidy.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy.go:903]: Should the same logic be added to `doTidyRevocationStore` as well I wonder?

I _think_ so, so perhaps we should wait until #677 lands to rebase on top of that (so that we correctly handle tidying each individually)....

Though I don't think we should add the `config.InvalidCerts` as an or to `if config.RevokedCerts || config.IssuerAssocs {` as we're already doing one iteration over the certs in this pass. That means we might miss certs (if for some reason, we have a revocation entry but no actual cert), but that should be fine.

My 2c!

Comment (fatima2003) [reply builtin/logical/pki/path_tidy.go:903]: I think that would be a good idea, maybe in the future we can have the option for `- verbose` in the case that a user wants to debug with details for whatever reason.

Comment (fatima2003) [reply builtin/logical/pki/path_tidy.go:903]: @cipherboy I will wait for #677 to land before I rebase this PR and modify the logging style in `doTidyRevocationStore`, then after this PR lands, I will open up the draft PR for using pagination in tidy and rebase there :D

**Comment by cipherboy** [comment]
@fatima2003 Do you mind rebasing this and I'll try to take a look and get it into Beta? Thanks!

File: builtin/logical/pki/path_tidy.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy.go:1088]: On new line 1100 is where I think I expected the `config.InvalidCerts` check, with additional warning if the cert copy in the revocation entry was of a bad certificate.

File: builtin/logical/pki/path_tidy.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy.go:825]: That also means I think on new line 836, you should also run the revocation store tidy if you are supposed to delete invalid certificates. They should be deleted from both places, I feel? My 2c!

**Review by cipherboy — COMMENTED** [review]
Hey Fatima, sorry about the delay! Two more minor things, it seems like I had forgotten to finish describing what I had meant by revocation should also warn + tidy invalid certs :-) 

Let me know what you think!

File: website/content/api-docs/secret/pki.mdx
Comment (cipherboy) [inline website/content/api-docs/secret/pki.mdx:4472]:
```suggestion
- `tidy_revoked_certs` `(bool: false)` - Set to true to remove all revoked 
  and expired certificates from storage. A revoked storage entry is considered
```

I think is what I actually meant to write here :D

File: website/content/api-docs/secret/pki.mdx
Comment (cipherboy) [inline website/content/api-docs/secret/pki.mdx:4479]: nit:

```suggestion
  from storage. A certificate is considered invalid if it cannot be parsed by Go's
  X.509 certificate parser.
```

File: website/content/api-docs/secret/pki.mdx
Comment (cipherboy) [inline website/content/api-docs/secret/pki.mdx:4643]:
```suggestion
    "tidy_revoked_certs": false,
```

File: builtin/logical/pki/path_tidy_test.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy_test.go:1343]: Nice! :-)

File: builtin/logical/pki/path_tidy_test.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy_test.go:1345]: While technically this works, I think this is invalid for an incorrect reason (wrong format) -- I think ideally you'd remove the `----BEGIN ... ` and `---END ....` headers and just have the base64 data and call `base64.StdEncoding.DecodeString(...)` on the body, since storage entries are the raw byte value of the DER form of the certificate (not the PEM-encoded "nice-to-read" form).

File: builtin/logical/pki/path_tidy_test.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy_test.go:1374]: I think you could reuse `waitForAutoTidyToFinish(...)` :-)

Comment (fatima2003) [reply builtin/logical/pki/path_tidy_test.go:1374]: Hmm, but I don't have a client defined, should I refactor the test to use clusters instead?

Comment (cipherboy) [reply builtin/logical/pki/path_tidy_test.go:1374]: Ahhh, I see. In that case, why not just invoke the relevant tidy method directly? Just a thought.

Comment (fatima2003) [reply builtin/logical/pki/path_tidy_test.go]: Oh wait, yea it is redundant

File: builtin/logical/pki/path_tidy_test.go
Comment (fatima2003) [reply builtin/logical/pki/path_tidy_test.go]: Good catch :) I missed this even though the error message states this 🙇🏽‍♀️

File: builtin/logical/pki/path_tidy_test.go
Comment (cipherboy) [inline builtin/logical/pki/path_tidy_test.go:1361]: @fatima2003 Sorry, I realized I was being confusing. I meant calling `doTidyCertStore` or `doTidyRevocationStore` with an appropriate `tidyConfig` object. It isn't fully realistic insofar as it isn't an end-to-end test, but for the purposes of what we're doing, I think it is fine!

This way, you don't even need the `time.Sleep` after this, as it is happening in the main thread and not in a background goroutine which may be interrupted.

Comment (fatima2003) [reply builtin/logical/pki/path_tidy_test.go]: Ahh okay, I didn't think that was an option. I pushed the update :)

**Review by cipherboy — APPROVED** [review]
Looks good, thanks @fatima2003!

**Comment by cipherboy** [comment]
Race detector failures unrelated, merging...

**Pull request merged**
